Use consistent layout for subforms in job/template forms

vjanssens opened the pull request:
##### SUMMARY
This PR resolves #5929 by moving the related form fields in job template create/edit and workflow template create/edit forms inside a subform. This changes the layout to be consistent with the rest of the application.

I could not reproduce some of the issues mentioned in #5929, so I guess they are alread resolved.


- checkbox fields should put options into 3 columns (like templates) -> was already resolved
- any dynamic fields based on checkbox options should be in gray/inset subform (like projects), and should have their fields in the standard 3 columns. -> fixed by this PR
- seems like whatever fields are supposed to show up on project form when update revision on launch is checked are not. -> was already resolved


##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
UI

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
12.0.0
```

##### ADDITIONAL INFORMATION

Job template form before:
![5929-job-template-before](https://user-images.githubusercontent.com/2340997/85038203-9c0c3080-b186-11ea-9eb1-d8c1fd9c4e53.png)

Job template form after:
![5929-job-template-after](https://user-images.githubusercontent.com/2340997/85038199-9adb0380-b186-11ea-9738-001e152cf0ca.png)

Workflow template form before
![5929-workflow-template-before](https://user-images.githubusercontent.com/2340997/85038209-9dd5f400-b186-11ea-9818-0127cbb8049d.png)

Workflow template form after:
![5929-workflow-template-after](https://user-images.githubusercontent.com/2340997/85038207-9d3d5d80-b186-11ea-973e-2a59f4d30a21.png)

vjanssens committed:
Use consistent layout for subforms in job/template forms

Signed-off-by: Vadiem Janssens <info@vadiemjanssens.nl>

jlmitch5 commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -175,39 +182,50 @@ function WorkflowJobTemplateForm({
+      </FormGroup>
+
+      {enableWebhooks && (
+        <>
```
you can get rid of the <> fragment cotainer here, only needed for when there are > 1 child.

jlmitch5 commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -175,39 +182,50 @@ function WorkflowJobTemplateForm({
+      {enableWebhooks && (
+        <>
+          <SubFormLayout>
+            <Title size="md">{i18n._(t`Webhook details`)}</Title>
```
can you add headingLevel="4" prop to Title here?  It's required in the patternfly v4 upgrade that will be merging shortly.

jlmitch5 commented on awx/ui_next/src/screens/Template/shared/JobTemplateForm.jsx:
```diff
@@ -543,30 +545,53 @@ function JobTemplateForm({
+                        />
+                      </FormColumnLayout>
+
+                      <br />
```
can you move the br out and put it between the two subforms?

```{allowCallbacks && enableWebhooks && (<br />)}```

jlmitch5 reviewed (changes requested):
looking good, thanks for the contribution!  a few little things

vjanssens committed:
Add headingLevels to Title, minor improvements

Signed-off-by: Vadiem Janssens <info@vadiemjanssens.nl>

vjanssens commented:
Thanks @jlmitch5 for the review, i've created a commit to fix the issues. 

Unfortunately, there's a test failing which I'm unable to fix. 
```
Warning: An update to WebhookSubForm inside a test was not wrapped in act(...).
When testing, code that causes React state updates should be wrapped into act(...):

<WorkflowJobTemplateForm/> › webhooks and enable concurrent jobs functions properly

    Method “props” is meant to be run on 1 node. 0 found instead.

      207 |     ).toBe(true);
      208 |     expect(
    > 209 |       wrapper.find('input[aria-label="wfjt-webhook-key"]').prop('readOnly')
          |                                                            ^
      210 |     ).toBe(true);
      211 |     expect(
      212 |       wrapper.find('input[aria-label="wfjt-webhook-key"]').prop('value')
```

This is my first time using react, but this error seems a bit vague. I haven't touched code inside WebookSubForm, why would it be changed? Can you point me in the right direction?
When I check the checkbox manually, I can see the `wfjt-webhook-key` be visible.

mabashian commented:
@vjanssens I see a couple of failing tests.  One of them has to do with webhook_key and webhook_url not matching our expected values on form submission.  I took a look at this and those fields are not even valid when we POST/PUT on `/api/v2/job_templates` so I don't think they should be included in the payload.  I would fix that test with the following changes:

https://github.com/mabashian/awx/commit/56ea5aa9070962d77edbd8bdef9c0da1a3ccb042

The other failing test is a bit more complicated.  I stopped the test right before the failing line and took a look at the DOM.  The `ContentLoading` component is being shown inside the subform which means that isLoading is still `true` when we try to access the field in the test.  While `ContentLoading` is shown, those fields won't be available and the test will fail.  The way that webhook subform worked before was that it was mounted as soon as the form mounted.  Now that you've wrapped it in a conditional render I think things are getting a bit sideways.  We may have to rethink the approach of the subform component in order to get this working smoothly.  For example, if we're going to mount the component when `enableWebhooks` is true then we don't need that same check inside the webhook subform component.  It will always be true if the component is mounted.

vjanssens committed:
Remove webhook_key and webhook_url from JT form payload

Signed-off-by: Vadiem Janssens <info@vadiemjanssens.nl>

vjanssens commented:
Thank you @mabashian. I've applied the fixes you proposed and removed the `enableWebhooks` conditional inside the WebhookSubForm, as I agree that double check was a bit strange.
I don't know the reasoning for why that state was passed to the component, but to me it is more straight forward to let the parent decide if it should be shown.

Fixed one of the tests, i'll need some more time to see if I can fix the other as well.

vjanssens committed:
Remove enableWebhooks conditional in WebhookSubForm

Signed-off-by: Vadiem Janssens <info@vadiemjanssens.nl>

vjanssens committed:
Fix failing test

Signed-off-by: Vadiem Janssens <info@vadiemjanssens.nl>

vjanssens commented:
recheck

vjanssens commented:
@mabashian I think I fixed the issue, I don't have any failing tests locally right now. However, the build seems to fail on something unrelated to my changes. Would you mind to take a look?

vjanssens commented:
Still not sure why the build fails. It does not seem to be related to my changes.

jlmitch5 commented:
@vjanssens I just force pushed your branch rebased on the latest devel branch (no other changes)...I assume something was just messed up on the check.  I pulled down the code and all tests pass and the UI looks good.  I assume it was something about how we've just merged a PR related to the UI packaging.

I'm putting the mergeit label on there, so provided the checks pass, it will merge.

Thanks for the contribution!

jlmitch5 reviewed (approved):
looks good!
